Use new PoPS Model API

malawrim opened the pull request:
Removed references to simulation.hpp and started integrating the new model class.

wenzeslaus commented on src/pops.cpp:
```diff
@@ -127,37 +127,20 @@ List pops_model(int random_seed,
-  Config config;
+  // TODO no establishment_probability, establishment_stochasticity, or establishment_probability
+  // steps and use_treatments set later (set to 0 and false for now)
+  Config config = {random_seed, num_rows, num_cols, ew_res, ns_res, 0, true, true, 0.5, 
+  use_lethal_temperature, lethal_temperature, weather, reproductive_rate, model_type_, latency_period,
+  natural_kernel_type, natural_distance_scale, natural_dir, natural_kappa, use_anthropogenic_kernel,
+  percent_natural_dispersal, anthropogenic_kernel_type, anthropogenic_distance_scale, anthropogenic_dir,
+  anthropogenic_kappa, false, mortality_on, mortality_rate, mortality_time_lag};
```
For better or worse, this won't be possible in the future with Config. Unlike Python, we can make clear key=value assignments in this context (at least with current and foreseeable C++) which makes it error prone (we still need to use these long lists of parameters elsewhere, but well...). Additionally, we will have to use more setters and getters in the Config class to keep some of the scheduling in consistent state. My original idea was that Config will be just a simple struct, but it will have to be a more complex class. However, the scheduling and configuration are interdependent, so my current solution for that is constructing Config in phases. Fixing that in the future may simplify Config class, but it won't fix problems with a long parameter list.

malawrim replied:
Just to clarify, for now I need to initialize each of the public variables separately (like I had it before) and then change these to calling setters when this functionality is expanded? Would it be possible to add a non-generic constructor to Config to avoid setting so many variables directly in pops.cpp?

wenzeslaus replied:
The answer to your first question is yes. I don't understand the second one. Do you want to prepare some pseudo-code of what you are envisioning? There should be a constructor with minimalistic defaults, so that the code does not crash when these are used, but all variables need to be set anyway because user needs to be able to set them, so I guess that's not what you mean.

malawrim replied:
I was thinking that instead of having a generic constructor for the config class [i.e Config() {...}] it could have a constructor with parameters that are currently the public variables in the config class [i.e Config(int random_seed, int rows, int cols, double ew_res, double ns_res, bool generate_stochasticity, .... ){...}] that way in the pops.cpp class instead of having ~20 lines of setting variables (i.e config.natural_kernel_type = natural_kernel_type; and so on) it would just call a constructor [Config config(seed, row, col, ew_res, ...)]. But, I was looking over the r.pops.spread main.cpp and I see that that approach does not work as well there.

wenzeslaus replied:
Right, it does not, but that is only part of the issue. The other part is that it is hard to get a long list of parameters right in C++ since you can't use their names and everything relies on order and type, but if you have a lot of variables with the same type, then it is really just the order. It is also less readable overall. Reducing number of lines would be nice, but not in this way. Compare readability and potential for errors in these four cases:

```
config1 = Config(temperature, year, month, day, num_steps);  // let's say this is correct and fits the def

config2 = Config();  // we can see right here that this is correct
config2.temperature = temperature;
config2.year = year;
config2.month = month;
config2.day = day;
config2.num_steps = num_steps;

config3 = Config(temperature, num_steps, year, month, day);  // oops, wrong order 

config4 = Config();
config4.temperature = num_steps;
config4.year = year;
config4.month = month;
config4.day = day;
config4.num_steps = temperature;  // pretty clear I made a copy-paste mistake here
```

I originally aimed at API with these public member assignments, but that won't work for things around scheduling and time. Setters are next closest solution, so we will move to those, although they have shortcomings such as longer code (not only lines, but also characters) and need for different states and checks in the object not needed if there is only a constructor (and thus one stage (only) initialization).

There are other ways in C++ how to solve this, but they are much less common. In Python and future C++20, there is a way of using key=value syntax in the constructor, but that's not available to us yet, I'm not sure if it would be actually useful as it would not address all our issues, and it is often written line-by-line anyway (e.g. with Python's Black formatting), so resulting in the same number of lines on the caller side.

This basically goes into review of https://github.com/ncsu-landscape-dynamics/PoPS/pull/70, so thanks for starting that already.

malawrim replied:
Thank you for the explanation, I understand now.

wenzeslaus commented on src/pops.cpp:
```diff
@@ -124,177 +125,145 @@ List pops_model(int random_seed,
-  SpreadRate<IntegerMatrix> spreadrate(infected, ew_res, ns_res, spread_rate_outputs);
-  // Define movement schedule  
+
+  unsigned count_lethal = config.num_lethal();
```
num_lethal() can be only called when lethal temperature is in use:

```
unsigned num_lethal()
    {
        if (!use_lethal_temperature)
            throw std::logic_error(
                "num_lethal() not available when use_lethal_temperature is false");
        if (!schedules_created_)
            throw std::logic_error(
                "Schedules were not created before calling num_lethal()");
        return get_number_of_scheduled_actions(lethal_schedule_);
    }
```

malawrim replied:
<img width="494" alt="Screen Shot 2020-07-16 at 11 42 40 AM" src="https://user-images.githubusercontent.com/63056769/87692287-92f88a00-c759-11ea-8933-56a64f45f989.png">

malawrim commented:
Only remaining issues are in test-abc_calibration and one last problem in test-pops. I believe the issues are due to model type SEI vs SI, but have not found the problem yet. Four warnings are due to parameter ordering in Radial_kernel.hpp, however these warnings do not appear to impact the results (tests errors are the same with/without the warnings fixed).

ChrisJones687 commented on src/pops.cpp:
```diff
@@ -45,269 +44,226 @@ bool all_infected(IntegerMatrix susceptible)
-)
-{
-  
+List pops_model(
```
I would like to move function calls to have a single variable on each line. I started doing this with the R functions but haven't done that here yet. It made it much easier to view changes.

ChrisJones687 commented on src/pops.cpp:
```diff
@@ -45,269 +44,226 @@ bool all_infected(IntegerMatrix susceptible)
-  
+      _["infected"] = infected_vector, _["exposed"] = exposed,
+      _["susceptible"] = susceptible_vector, _["resistant"] = resistant_vector,
+      _["mortality"] = mortality_vector, _["rates"] = spread_rates_vector,
```
I would prefer each export variable to be on a separate line, it makes it easier to glance and see what we are exporting when we want to add in new functionality like the quarantine.

malawrim commented:
> 1. The `abc_calibration` is still commented out. Is that an intention?

I removed the comments from the first test because I added those, but I am not sure about the second one. @ChrisJones687 would you like me to uncomment the second abc_calibration test?

> 2. The clang-format style applied is the default one. For PoPS C++ lib I used a custom one (`.clang-format` file). Are you aware of that? I'm not saying we have to use it here, but I was not sure if you used the default clang-format style because it more fits with what we have in R code or if there was other reason. (The one in C++ lib is made to resamble Python PEP8 and Black styles, but it is not perfect anyway.)

I just used the default one because I was not sure if there was any specific format to use. I can change it to match the PoPS library if you would like.

wenzeslaus commented:
> > 1. The clang-format style...
> 
> I just used the default one because I was not sure if there was any specific format to use. I can change it to match the PoPS library if you would like.

Given that our other projects are in Python and there is no single strong formatting practice in R (or in C++), I would say, let's use the one from PoPS. You need to copy the file `.clang-format` from PoPS to the root directory of this repo or perhaps the directory with the C++ source code in this repo makes more sense. clang-format should pick that file automatically (if not `--style=file` should trigger that behavior).

ChrisJones687 commented:
This is good on my end. Merge once @wenzeslaus approves clang-format. Once merged we can add the new quarantine metrics as outputs and bugfix.

ChrisJones687 commented:
We need to .clang-format to .Rbuildignore and it should fix the building on appveyor issue
